File: src/key.ts
import * as crypto from 'crypto';
import baseX from 'base-x';

export const bs58 = baseX('123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz');

interface Algorithm {
  encrypt(message: Buffer, addition: Buffer | undefined, nonce: Buffer): Buffer;
  decrypt(message: Buffer, additionLength: number, nonce: Buffer): { dec: Buffer; additional?: Buffer };
}

export type supportedAlgs = 'chacha20-poly1305';

export class ChaCha20Poly1305 implements Algorithm {
  readonly opts: crypto.CipherCCMOptions = {
    authTagLength: 4,
  };
  readonly key: Buffer;
  readonly alg: crypto.CipherCCMTypes = 'chacha20-poly1305';

  constructor(key: Buffer, opts?: crypto.CipherCCMOptions) {
    this.key = key;
    if (opts) {
      this.opts = opts;
    }
  }

  encrypt(message: Buffer, addition: Buffer | undefined, nonce: Buffer): Buffer {
    const cipher = crypto.createCipheriv(this.alg, this.key, nonce, this.opts);
    // if (addition) {
      // cipher.setAAD(addition, { plaintextLength: addition?.length });
    // }
    const ret = cipher.update(message);
    cipher.final();
    return ret;
  }
  decrypt(dec: Buffer, additionLength: number, nonce: Buffer): { dec: Buffer; additional?: Buffer } {
    const decipher = crypto.createDecipheriv(this.alg, this.key, nonce, this.opts);
    const ret = decipher.update(dec);
    // decipher.setAAD(Buffer.alloc(additionLength));
    // read additional
    // decipher.
    decipher.final();
    return { dec: ret };
  }

}

export function createAlgorithm(alg: supportedAlgs, key: Buffer): Algorithm {
  if (alg === 'chacha20-poly1305') {
    return new ChaCha20Poly1305(key);
  }
  throw new Error('Invalid or unsupported algorithm');
}

export interface SymetricKeyProp {
  readonly id?: string;
  readonly hashSeed?: string;
  readonly key?: string;
  readonly nonce?: string;
  readonly alg?: crypto.CipherCCMTypes;
  readonly keyLength?: number;
  readonly nonceLength?: number;
}

export class SymetricKey {
  id?: string;
  readonly hashSeed: string;
  keyBuffer?: Buffer;
  nonceBuffer?: Buffer;
  readonly alg: crypto.CipherCCMTypes;
  readonly keyLength: number;
  readonly nonceLength: number;

  constructor(prop: SymetricKeyProp) {
    this.id = prop.id;
    this.hashSeed = prop.hashSeed || crypto.randomUUID();
    this.alg = prop.alg || 'chacha20-poly1305';
    this.keyBuffer = prop.key ? Buffer.from(prop.key, 'base64') : undefined;
    this.nonceBuffer = prop.nonce ? Buffer.from(prop.nonce, 'base64') : undefined;
    this.keyLength = prop.keyLength || 32;
    this.nonceLength = prop.nonceLength || 12;
  }

  public static async create(props: SymetricKeyProp): Promise<SymetricKey> {
    return new SymetricKey(props).create();
  }

  async create(): Promise<SymetricKey> {
    return new Promise<SymetricKey>((rs, rj) => {
      crypto.randomBytes(this.keyLength, (err, key) => {
        if (err) {
          rj(err);
          return;
        }
        crypto.randomBytes(this.nonceLength, (err, nonce) => {
          if (err) {
            rj(err);
            return;
          }
          this.keyBuffer = key;
          this.nonceBuffer = nonce;
          const hash = crypto.createHash('sha256');
          hash.update(key);
          hash.update(nonce);
          this.id = hash.digest().toString('base64');
          rs(this);
        });
      });
    });
  }

  get key(): string {
    if (!this.keyBuffer) {
      throw Error('key is not set!');
    }
    const key = this.keyBuffer.toString('base64') as string;
    return key;
  }

  get nonce(): string {
    if (!this.nonceBuffer) {
      throw Error('nonce is not set!');
    }
    const nonce = this.nonceBuffer.toString('base64') as string;
    return nonce;
  }

  decrypt(dec: string): Buffer {
    if (!dec.startsWith('C')) {
      throw Error(`Unknown cipher:${dec}`);
    }
    dec = dec.slice('C'.length);
    const decipher = crypto.createDecipheriv(
      this.alg,
      Buffer.from(this.key, 'base64'),
      Buffer.from(this.nonce, 'base64'),
      {
        authTagLength: 4,
      },
    );
    const ret = decipher.update(bs58.decode(dec));
    decipher.final();
    return ret;
  }

  encrypt(enc: string | Buffer, addition?: Buffer): string {
    const cipher = crypto.createCipheriv(this.alg, Buffer.from(this.key, 'base64'), Buffer.from(this.nonce, 'base64'), {
      authTagLength: 4,
    });
    let out: Buffer;
    if (typeof enc === 'string') {
      out = Buffer.from(enc);
    } else {
      out = enc;
    }
    if (addition) {
      cipher.setAAD(addition, { plaintextLength: addition?.length });
    }
    const ret = cipher.update(out);
    cipher.final();
    // C means chacha20-poly1305
    return 'C' + bs58.encode(ret);
  }

  hash(enc: string): string {
    const hasher = crypto.createHash('sha256');
    hasher.update(this.hashSeed);
    hasher.update(enc.toUpperCase().replace(/[\n\r \t]+/g, ''));
    // 'A' means sha256 upcase und replace
    return 'A' + bs58.encode(hasher.digest());
  }
}

File: src/key.py
from dataclasses import dataclass
import uuid
import os
import re
import base58
from typing import Optional

from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
from cryptography.hazmat.primitives import hashes

@dataclass
class SymetricKeyProps:
    alg: str
    id: Optional[str] = None
    nonce: bytes = os.urandom(12)
    key: Optional[bytes] = None
    hashSeed: str = str(uuid.uuid4())
    key_length: int = 32
    nonce_length: int = 12

class SymetricKey:
    _id: Optional[str]
    _hashSeed: str
    _key: Optional[bytes]
    _nonce: bytes
    _key_length: int
    _nonce_length: int
    _alg = 'chacha20-poly1305'


    def __init__(self, props: SymetricKeyProps):
        self._id = props.id
        self._key = props.key
        self._hashSeed = props.hashSeed
        self._key_length = props.key_length
        self._nonce_length = props.nonce_length
        self._nonce = props.nonce

    @staticmethod
    def create(props: SymetricKeyProps):
        return SymetricKey(props).generate()

    # def generate_nonce(self, length: Optional[int] = None) -> bytes:
        # nonce_length = length if length else self._nonce_length
        # self._nonce = os.urandom(nonce_length)
        # return self._nonce

    def generate(self):
        if self._key is None:
            self._key = os.urandom(self._key_length)
        self._nonce = self._nonce if self._nonce else self.create_nonce()
        if self._id is None:
            digest = hashes.Hash(hashes.SHA256())
            digest.update(self._key)
            digest.update(self._nonce)
            self._id=base58.b58encode(digest.finalize()).decode()
        self._chacha = ChaCha20Poly1305(self._key)
        return self

    def create_nonce(self) -> bytes:
        self._nonce = os.urandom(12)
        return self.nonce

    @property
    def key(self) -> bytes:
        if self._key is None:
            raise Exception("key is not set")
        return self._key

    @property
    def id(self) -> str:
        if self._id is None:
            raise Exception("key is not set")
        return self._id

    @property
    def nonce(self) -> bytes:
        if self._nonce is None:
            raise Exception("nonce is not set")
        return self._nonce

    def encrypt(self, nonce: bytes, message: bytes, addition: Optional[bytes] = None ) -> bytes:
        return self._chacha.encrypt(nonce, message, addition)

    def decrypt(self, nonce: bytes, encrypted: bytes, addition: Optional[bytes] = None) -> bytes:
        return self._chacha.decrypt(nonce, encrypted, addition)

    def hash(self, enc: str)-> str:
        hasher = hashes.Hash(hashes.SHA256())
        hasher.update(bytes(self._hashSeed, 'utf-8'));
        stripped = re.sub(r'[\n\r\s\t]+','', enc.upper(),flags=(re.M))
        hasher.update(stripped.encode())
        # 'A' means sha256 upcase und replace
        string_hash = base58.b58encode(hasher.finalize()).decode()
        return  "A{}".format(string_hash)



